llvm/lib/Target/AIE/AIEHazardRecognizer.cpp

```diff
@@ -563,3 +603,18 @@ AIEHazardRecognizer::getSelectedAltOpcode(MachineInstr *MI) const {
     return It->second;
   return std::nullopt;
 }
+
+MemoryBankBits AIEHazardRecognizer::getMemoryBanks(MachineInstr *MI) const {
+  MemoryBankBits MemoryBankUsed = 0;
+
+  if (!(MI->mayLoad() || MI->mayStore()))
+    return MemoryBankUsed;
+
+  const AIEBaseSubtarget &STI = AIEBaseSubtarget::get(*MI->getMF());
+  for (auto &MMO : MI->memoperands()) {
+    MemoryBankBits MemoryBank =
+        STI.getMemoryBanksFromAddressSpace(MMO->getAddrSpace());
+    MemoryBankUsed |= MemoryBank;
```

Comment:
What happens for TM accesses? For those, I think we have two MMOs: one from the IR value, and a custom one to disambiguate DM from TM. If the IR value does not have the right address space, I guess we end up blocking all banks, even though we know from the second MMO that it's in the TM bank?

Reply:
We could maybe actually start with `MemoryBankUsed = ~0;`, and let every MMO refine the banks with `MemoryBankUsed &= MemoryBank` instead of `|=`. 

In the first commit, I think we should define `STI.getMemoryBanksFromPseudoSourceValue` as well. E.g.
```
getMemoryBanksFromPseudoSourceValue(const PseudoSourceValue &PSV) {
  switch (PSV.getKind()) {
  case TileMemory: return Bank::TileMemory
  default: return Bank::all()
  }
}
```

Then in the loop here, we can use getMemoryBanksFromPseudoSourceValue if `MMO.getPseudoValue()` returns a valid pointer.

Reply:
Updated as suggested.